[BE] user-interest one-to-many 매핑

bohyeon-n opened the pull request:
## user와 interest one-to-many 매핑

user와 interest를 one-to-many 관계를 가지도록 매핑하였습니다. 
Set 컬렉션을 사용해서 interest 를 매핑하려면 user_id 컬럼만 추가하면 되는데, 순서가 있는 List를 사용하기 위해서 user_key 컬럼을 추가하였습니다.  
interest에서 제일 관심있는 분야를 먼저 보여주면 좋을 것 같다고 생각했습니다. 

[보내주신 영상](https://www.youtube.com/watch?v=ccxBXDAPdmo)과 [data jdbc 문서](https://docs.spring.io/spring-data/jdbc/docs/1.1.0.RC2/reference/html/#jdbc.entity-persistence.types)를 참고했습니다. 

develop 으로 pr 보내기 전에 @jypthemiracle 작업하고 있던 브랜치에 pr 보냅니다!

realsigridjin commented on BE-workspace/src/main/resources/schema.sql:
```diff
@@ -1,8 +1,18 @@
 
 create table user (
-    id serial primary key,
+    id bigint auto_increment primary key,
```
bigint를 굳이 사용하신 이유가 있나요?

realsigridjin commented on BE-workspace/src/main/java/com/codesquad/signup4/domain/User.java:
```diff
@@ -10,22 +13,42 @@
-  User(Long id, String userId, String password, String email) {
-    this.id = id;
+
+  @MappedCollection(idColumn = "user_id", keyColumn = "user_key")
```
```The @MappedCollection annotation can be used on a reference type (one-to-one relationship) or on Sets, Lists, and Maps (one-to-many relationship).``` 저는 자료구조를 통해 ORM을 구성하는 것이 신기했어요. 따로 어노테이션이 없어서 relationship 관련 내용을 찾는 것이 어려웠구나, 공식문서를 잘 보는 것이 최고구나 라는 점을 다시 한 번 느꼈습니다. 학습하시면서 엘리는 어떠셨나요? 공식문서는 뭔가 예시가 풍부하지 않아서 전 겁이 났는데, 구현하신 것이 대단하다고 느꼈어요!

realsigridjin replied:
제가 이해하기에는 List와 Set에서 idColumn과 keyColumn을 별도로 갖는 이유가 uniqueness를 보장하기 위함, 즉 id값을 서로 구분해주기 위해 일종의 index(key이겠죠)를 만들어준다고 생각했습니다. 해당 이해가 맞을까요?

bohyeon-n replied:
저도 잘 모르겠지만, 제가 이해한 거는 idColumn은 외래 키의 이름을 지정해 주는 것이고, keyColumn은 List에서 해당 데이터의 인덱스(위치)를 위한 컬럼 이름을 정해주는 것이라고 생각했습니다! 그래서 Set 컬렉션을 사용할 때는 keyColumn을 지정하지 않아도 되더라구요!

realsigridjin commented on BE-workspace/src/main/java/com/codesquad/signup4/domain/User.java:
```diff
@@ -10,22 +13,42 @@
-    this.id = id;
+
+  @MappedCollection(idColumn = "user_id", keyColumn = "user_key")
+  private List<Interest> interest;
```
혹시 ```List```를 쓰신 이유가 있나요? 저는 중복 염려가 없으니 ```Set```이 더 적합하지 않나 생각했습니다. 굳이 ```KeyColumn```을 추가로 설정할 이유도 없구요.

realsigridjin replied:
아, PR에 이유가 적혀있네요. ```interest에서 제일 관심있는 분야를 먼저 보여주면 좋을 것 같다고 생각했다```고 하셨는데 어떤 말씀인지 더 알려주시면 좋을 것 같아요!

bohyeon-n replied:
관심사에 우선순위가 있다면, 사용자가 선택한 순서까지 데이터에 저장해주면 좋지 않을까 생각했습니다! 그런데 이건 나중에 생각한 거고, 영상 따라해보면서 왜 Set을 쓰는거지? 하다가 List로 똑같이 해봤는데 안되더라구요!

bohyeon-n replied:
근데 이건 바꿔도 될 것 같아요! 순서 없어도 상관 없을 것 같아요!

realsigridjin commented on BE-workspace/src/main/resources/schema.sql:
```diff
@@ -1,8 +1,18 @@
 
 create table user (
-    id serial primary key,
+    id bigint auto_increment primary key,
```
bigint를 굳이 사용하신 이유가 있나요?

bohyeon-n replied:
user에서 정한 id가 Long 타입이어서 bigint 를 사용하였습니다. 사실 serial이 뭔지 잘 몰랐어요...ㅎ 
serial은 auto_increment 일 때 연속되는 값에 사용한다고 하는데 bigserial도 있네요!

bohyeon-n replied:
저도 잘 모르겠지만, 제가 이해한 거는 idColumn은 외래 키의 이름을 지정해 주는 것이고, keyColumn은 List에서 해당 데이터의 인덱스(위치)를 위한 컬럼 이름을 정해주는 것이라고 생각했습니다! 그래서 Set 컬렉션을 사용할 때는 keyColumn을 지정하지 않아도 되더라구요!

bohyeon-n committed:
user-interest one-to-many 매핑

user와 interest one-to-many 관계로 매핑함
interest가 순서를 갖도록 하기 위해 List 컬렉션을 사용함, user_key 컬럼을 추가함
소스 코드만 올리기 위해서.gitignore 파일 추가하고 이미 올라와 있던 파일들 삭제함

bohyeon-n commented:
@jypthemiracle  .gitignore 파일이 없어서 추가하고 이미 올라와 있던 파일은 삭제하였습니다!

realsigridjin commented on BE-workspace/src/main/java/com/codesquad/signup4/domain/User.java:
```diff
@@ -10,22 +13,42 @@
     this.userId = userId;
     this.password = password;
     this.email = email;
+    this.interest = interest;
```
인자 값으로 interest를 받지 않는데, 혹시 ```this.interest = interest``` 가 어떤 의미가 있나요?

bohyeon-n replied:
앗... 수정하다가 못지운 코드에요... ㅎ

sigridjineth reviewed (commented):
1. 조금 이따 8시에 Zoom으로 함께 코드에 대해 이야기 나눠봐요.
2. BE/feature/17/data-mapping 브랜치 쓰셨는데 이거 좋은 것 같네요. 이슈번호/피쳐설명이지요?
3. 엘리 짱짱이에요! 오늘 수고 많으셨어요.

bohyeon-n committed:
keyColumn 컬럼 이름 interest_key로 수정함

bohyeon-n committed:
User 클래스에 필드 추가

누락된 필드 추가함, 코드 정렬함

bohyeon-n committed:
user 테이블 컬럼 추가함

나머지 user 데이터 컬럼 추가함
일단 varchar로 통일해놨는데 확인해주세요!

bohyeon-n commented:
@jypthemiracle  여기에 작업해버렸네요... 😂
이거 BE-dev 로 머지하면 될 것 같아요!
